### ISSUE TITLE
Bug: [switch-exhaustiveness-check] invalid fix for union with infinite types

### ISSUE THREAD

**Issue by Josh-Cena** [typescript-eslint/typescript-eslint#6682]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.0.1-rc&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPARFPvAD4EBGxZKArgLbkgwDcAUK2gO4CWGYAFvAAUUAJTwA3q3iIoaBIXx5ycKAGs2MsHIWVlqjawC+rIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6KImjQA9tEjgwAXxAGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
declare const a: "a" | "b" | number;

switch (a) {
  case "a": break;
  case "b": break;
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/switch-exhaustiveness-check": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No error, no suggestion fixer, or suggestion to add a default case

### Actual Result

Fixer produces `case number:`, which is invalid syntax.

### Additional Info

I'm not sure what the best fix for this issue is, so I hope to get more discussion first.

**Comment by Josh-Cena** [comment]
Found another related instance:

```ts
declare const a: number | string;

switch (a) {
  case 2: break;
}
```

This should not report anything

**Comment by Josh-Cena** [comment]
Another interesting case:

```ts
declare const a: "a" | (string & {});

switch (a) {
  case "a": break;
}
```

This has a suggestion but suggestion produces an empty line

**Comment by bradzacher** [comment]
Report if the literal cases are missing with no default, no auto or suggestion fixer provided.

Considering we ignore the `number` etc case normally - I'd say the rule should continue to ignore it?

**Comment by Josh-Cena** [comment]
That's reasonable. So we filter out all infinite types from the union, and require the remaining literals to be exhaustively matched.
